pkg/kubelet/volume/reconciler/reconciler.go

```diff
@@ -0,0 +1,231 @@
+//   controller is responsible for managing the attach/detach operations for
+//   this node, and therefore the volume manager should not
+// loopSleepDuration - the ammount of time the reconciler loop sleeps between
+//   successive executions
+// waitForAttachTimeout - the ammount of time the Mount function will wait for
+//   the volume to be attached
+// hostName - the hostname for this node, used by Attach and Detach methods
+// desiredStateOfWorld - cache containing the desired state of the world
+// actualStateOfWorld - cache cotnaining the actual state of the world
+// operationExecutor - used to trigger attach/detach/mount/unmount operations
+//   safely (prevents more than one operation from being triggered on the same
+//   volume)
+func NewReconciler(
+	controllerAttachDetachEnabled bool,
+	loopSleepDuration time.Duration,
+	waitForAttachTimeout time.Duration,
+	hostName string,
+	desiredStateOfWorld cache.DesiredStateOfWorld,
+	actualStateOfWorld cache.ActualStateOfWorld,
+	operationExecutor operationexecutor.OperationExecutor) Reconciler {
+	return &reconciler{
+		controllerAttachDetachEnabled: controllerAttachDetachEnabled,
+		loopSleepDuration:             loopSleepDuration,
+		waitForAttachTimeout:          waitForAttachTimeout,
+		hostName:                      hostName,
+		desiredStateOfWorld:           desiredStateOfWorld,
+		actualStateOfWorld:            actualStateOfWorld,
+		operationExecutor:             operationExecutor,
+	}
+}
+
+type reconciler struct {
+	controllerAttachDetachEnabled bool
+	loopSleepDuration             time.Duration
+	waitForAttachTimeout          time.Duration
+	hostName                      string
+	desiredStateOfWorld           cache.DesiredStateOfWorld
+	actualStateOfWorld            cache.ActualStateOfWorld
+	operationExecutor             operationexecutor.OperationExecutor
+}
+
+func (rc *reconciler) Run(stopCh <-chan struct{}) {
+	wait.Until(rc.reconciliationLoopFunc(), rc.loopSleepDuration, stopCh)
+}
+
+func (rc *reconciler) reconciliationLoopFunc() func() {
+	return func() {
+		// Unmounts are triggered before mounts so that a volume that was
+		// referenced by a pod that was deleted and is now referenced by another
+		// pod is unmounted from the first pod before being mounted to the new
+		// pod.
+
+		// Ensure volumes that should be unmounted are unmounted.
+		for _, mountedVolume := range rc.actualStateOfWorld.GetMountedVolumes() {
+			if !rc.desiredStateOfWorld.PodExistsInVolume(mountedVolume.PodName, mountedVolume.VolumeName) {
+				// Volume is mounted, unmount it
+				glog.V(12).Infof("Attempting to start UnmountVolume for volume %q (spec.Name: %q) from pod %q (UID: %q).",
+					mountedVolume.VolumeName,
+					mountedVolume.OuterVolumeSpecName,
+					mountedVolume.PodName,
+					mountedVolume.PodUID)
+				err := rc.operationExecutor.UnmountVolume(
+					mountedVolume.MountedVolume, rc.actualStateOfWorld)
+				if err == nil {
```

Comment:
IsAlreadyExists errors are expected and common, hence not logging them. Added comment explaining.